File: app/src/main/java/org/fundacionparaguaya/adviserplatform/ui/common/widget/NumberStepperView.java
package org.fundacionparaguaya.adviserplatform.ui.common.widget;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.content.ContextCompat;
import android.support.v4.widget.TextViewCompat;
import android.support.v7.widget.AppCompatImageButton;
import android.support.v7.widget.AppCompatTextView;
import android.util.AttributeSet;
import android.view.View;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.FrameLayout;
import android.widget.TextSwitcher;
import org.fundacionparaguaya.adviserassistant.R;

public final class NumberStepperView extends FrameLayout {

    private static final int DESIRED_HEIGHT = 178;
    private static final int DESIRED_WIDTH = 56;
    private int minValue = Integer.MIN_VALUE;
    private int maxValue = Integer.MAX_VALUE;

    private static final float DISABLED_ALPHA = 0.2f;
    private static final float ENABLED_ALPHA = 1f;

    private final AppCompatImageButton mMinusButton;
    private final AppCompatImageButton mPlusButton;

    private int currentValue;
    private final TextSwitcher mTextSwitcher;
    
    private final MutableLiveData<Integer> valueLiveData = new MutableLiveData<>();

    public final void setMinValue(int value) {
        minValue = value;
        if(currentValue < value) {
            setCurrentValue(value);
        }
    }

    public final void setMaxValue(int value) {
        maxValue = value;
    }

    public final void setCurrentValue(int i) {
        if(i >= minValue || i <= maxValue) {
            currentValue = i;
            mTextSwitcher.setText(Integer.toString(currentValue));
        }

        valueLiveData.setValue(currentValue);
        updateButtons();
    }

    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        this.setMeasuredDimension(DESIRED_WIDTH, DESIRED_HEIGHT);
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
    }

    private void addListeners() {
        this.mPlusButton.setOnClickListener(v -> setCurrentValue(currentValue+1));
        this.mMinusButton.setOnClickListener(v -> setCurrentValue(currentValue-1));
    }

    private void updateButtons() {
        setButtonEnabled(mMinusButton, currentValue > minValue);
        setButtonEnabled(mPlusButton, currentValue < maxValue);
    }

    private void setButtonEnabled(AppCompatImageButton button, boolean enabled) {
        button.setEnabled(enabled);
        if(button.isEnabled()) {
            button.setAlpha(ENABLED_ALPHA);
        } else {
            button.setAlpha(DISABLED_ALPHA);
        }

    }

    public final LiveData<Integer> getValue() {
        return this.valueLiveData;
    }

    public NumberStepperView(@NonNull final Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);

        View view = View.inflate(context, R.layout.view_numberstepper, null);

        mMinusButton = view.findViewById(R.id.btn_numberstepper_minus);
        mPlusButton = view.findViewById(R.id.btn_numberstepper_plus);
        mTextSwitcher = view.findViewById(R.id.textswitcher_numberstepper);

        mTextSwitcher.setFactory((() -> {
            AppCompatTextView textView = new AppCompatTextView(context, null, 0);
            TextViewCompat.setTextAppearance(textView, R.style.BigHero);
            textView.setTextColor(ContextCompat.getColor(context, R.color.app_white));
            return textView;
        }));

        Animation inAnim = AnimationUtils.loadAnimation(context, android.R.anim.fade_in);
        Animation outAnim = AnimationUtils.loadAnimation(context, android.R.anim.fade_out);
        inAnim.setDuration(200);
        outAnim.setDuration(200);

        mTextSwitcher.setInAnimation(inAnim);
        mTextSwitcher.setOutAnimation(outAnim);

        addView(view);
        addListeners();

        if(isInEditMode()) {
            mTextSwitcher.setText("2");
        }
    }
}

File: app/src/main/java/org/fundacionparaguaya/advisorapp/viewcomponents/NumberStepperView.kt
package org.fundacionparaguaya.advisorapp.viewcomponents

import android.arch.lifecycle.MutableLiveData
import android.content.Context
import android.support.v4.content.ContextCompat
import android.support.v4.widget.TextViewCompat
import android.support.v7.widget.AppCompatImageButton
import android.support.v7.widget.AppCompatTextView
import android.util.AttributeSet
import android.view.View
import android.view.animation.AnimationUtils
import android.widget.FrameLayout
import android.widget.TextSwitcher
import org.fundacionparaguaya.advisorapp.R

/**
 * Custom compound view for a number stepper/picker. Provides two buttons, a minus and a plus button, that is used to
 * increment or decrement the value.
 */

class NumberStepperView(context: Context, attrs: AttributeSet?) : FrameLayout(context, attrs) {

    val desiredHeight = 178
    val desiredWidth = 56

    var minValue: Int = Int.MIN_VALUE
    set(i) //this set function is triggered whenever minValue = [some value] happens..
    {
        field = i //this sets the actual value of the field

        if(currentValue<i)
        {
            currentValue = i
        }
    }

    var maxValue: Int = Int.MAX_VALUE

    private val DISABLED_ALPHA = 0.2f
    private val ENABLED_ALPHA = 1f

    private val mMinusButton: AppCompatImageButton
    private val mPlusButton: AppCompatImageButton
    var valueLiveData = MutableLiveData<Int>()

    var currentValue: Int = 0
    set(i)
    {
        if(i >= minValue || i <= maxValue)
        {
            field = i
            mTextSwitcher.setText(Integer.toString(i))
        }

        valueLiveData.value = i

        updateButtons()
    }

    /* Text switchers allows for nicer animations between values, particularly double digit values */
    private val mTextSwitcher: TextSwitcher

    init {

        val view = View.inflate(context, R.layout.view_numberstepper, null)

        mMinusButton = view.findViewById(R.id.btn_numberstepper_minus)
        mPlusButton = view.findViewById(R.id.btn_numberstepper_plus)
        mTextSwitcher = view.findViewById(R.id.textswitcher_numberstepper)

        mTextSwitcher.setFactory {
            val textView = AppCompatTextView(context, null, 0)
            TextViewCompat.setTextAppearance(textView, R.style.BigHero)
            textView.setTextColor(ContextCompat.getColor(context, R.color.app_white))

            return@setFactory textView
        }

        val inAnim = AnimationUtils.loadAnimation(context,
                android.R.anim.fade_in)

        val outAnim = AnimationUtils.loadAnimation(context,
                android.R.anim.fade_out)

        inAnim.duration = 200
        outAnim.duration = 200

        mTextSwitcher.inAnimation = inAnim
        mTextSwitcher.outAnimation = outAnim

        addView(view)
        addListeners()

        if(isInEditMode) //this is just used to provide default text in the editor preview
        {
            mTextSwitcher.setText("2")
        }
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {

        setMeasuredDimension(desiredWidth, desiredHeight)
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
    }

    private fun addListeners()
    {
        mPlusButton.setOnClickListener({ _ -> currentValue++})
        mMinusButton.setOnClickListener({ _ -> currentValue--})
    }

    private fun updateButtons()
    {
        setButtonEnabled(mMinusButton, currentValue >= minValue)
        setButtonEnabled(mPlusButton, currentValue <= maxValue)
    }

    /**
     * Sets a button (either plus or minus) to enabled/disabled depending on whether the max/min has been reached.
     * Also dims the button if it is disabled
     */
    private fun setButtonEnabled(button: AppCompatImageButton, enabled: Boolean)
    {
        button.isEnabled = enabled

        if(button.isEnabled)
        {
            button.alpha = ENABLED_ALPHA
        }
        else
        {
            button.alpha = DISABLED_ALPHA
        }
    }

    /**
     * Returns a live data object that can be observed to hear changes in this component's value
     */
    fun getValue() = valueLiveData
}
